### PR TITLE
Add encryption version fix command

### PR DESCRIPTION
**Pull request by sharidas**
Add a new command to fix the encryption
version. This would help the admin to
fix the issues related to encryption version
mismatch for the files.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

### CODE REVIEW

**Comment by sharidas** [comment]
- [ ] Improve the script
- [ ] Add unit tests

**Comment by sharidas** [comment]
This error https://drone.owncloud.com/owncloud/encryption/599/150 is not reproducible in my local instance. Digging into it...

**Comment by sharidas** [comment]
Using the unit test this is what I wanted to do:
- Simulate the scenario that happened with one of our user(s), where the filecache table got messed up with wrong encrypted version
- Run the command
- Verify everything is ok

But the problem here I find with
#### Unit test
- Simulating and executing the command is as shown in the https://github.com/owncloud/encryption/pull/115#issuecomment-479895755
- The drone failure https://drone.owncloud.com/owncloud/encryption/599/150.  Especially `openssl_seal(): not a public key (1th member of pubkeys)`
I am thinking of skipping this test for now in the unit test.

#### Acceptance test
- As of now we don't have the end point in the tests where the db could be manipulated. Once its there we could use it to alter the values and run the command and check the files. IMHO, that would be a clean way.

File: lib/Command/FixEncryptedVersion.php [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+						$output->writeln("<info>Fixed the file $path with version " . $encryptedVersion . "</info>");
+						return true;
+					}
+					$encryptedVersion--;
+				}
+			}
+
+			//So decrementing did not worked. Now lets increment. Max increment is till 25
+			$increment = 1;
+			while ($increment <= 25) {
```
Comment (PVince81) [inline lib/Command/FixEncryptedVersion.php]: not sure if we need to go that far, this would also take more time to process for many files

I suspect that the root cause would only cause a discrepancy of +1.
maybe use 5 here just to be safe.

File: lib/Command/FixEncryptedVersion.php [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+				if ($oldEncryptedVersion !== null) {
```
Comment (PVince81) [inline lib/Command/FixEncryptedVersion.php]: why not directly assign `$oldEncryptedVersion` to `$encryptedVersion` at the beginning? unnecessary extra logic makes it more confusing and less readable

File: lib/Command/FixEncryptedVersion.php [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+					$encryptedVersion = $oldEncryptedVersion + $increment;
+				} else {
+					$encryptedVersion = $encryptedVersion + $increment;
```
Comment (PVince81) [inline lib/Command/FixEncryptedVersion.php]: please recheck, from my understanding this will go as follows:

with $encryptedVersion = 6 at start
 
cycle 1: $increment=1, $encryptedVersion=6+1=7
cycle 2: $increment=2, $encryptedVersion=7+2=9
cycle 3: $increment=3, $encryptedVersion=9+3=12
...

if it doesn't do that (which might be confirmed with unit test), then rewrite the code to make it more clear what happens to the reader

File: tests/unit/Command/FixEncryptedVersionTest.php [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+		$this->commandTester->execute([
+			'user' => self::TEST_ENCRYPTION_VERSION_AFFECTED_USER
+		]);
+
+		$output = $this->commandTester->getDisplay();
+		$this->assertEquals("Verifying the content of file /test_enc_version_affected_user1/files/hello.txt
+Decrement the encrypted version to 1
+Increment the encrypted version to 3
+Increment the encrypted version to 4
```
Comment (PVince81) [inline tests/unit/Command/FixEncryptedVersionTest.php]: please test with further increments as I seem to have spotted a bug in the increment logic... see my other comment above

File: lib/Command/FixEncryptedVersion.php [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+	}
+
+	protected function configure() {
+		parent::configure();
+
+		$this
+			->setName('encryption:fixencryptedversion')
+			->setDescription('Fix the encrypted version if the encrypted file(s) are not downloadable from the UI')
+			->addArgument(
+				'user',
```
[the remark this answers is not in the document]
Comment (sharidas) [reply lib/Command/FixEncryptedVersion.php]: I was thinking of adding that change later to the script.

File: lib/Command/FixEncryptedVersion.php [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+					$this->verifyFileContent($path, $output);
+				}
+			}
+		}
+	}
+
+	/**
+	 * @param string $path
+	 * @param OutputInterface $output
+	 * @param bool $allowOnce
```
[the remark this answers is not in the document]
Comment (sharidas) [reply lib/Command/FixEncryptedVersion.php]: Renamed this to `$ignoreCorrectEncVersionCall`. Idea is not to call 
```
$this->correctEncryptedVersion()
```
in the `catch` portion of the code.

File: lib/Command/FixEncryptedVersion.php [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+	 */
+	private function verifyFileContent($path, OutputInterface $output, $allowOnce = true) {
+		try {
+			if ($this->view->readfilePart($path, 0, 9000) !== false) {
+				$output->writeln("<info>The file $path is: OK</info>");
+			}
+			return true;
+		} catch (HintException $e) {
```
[the remark this answers is not in the document]
Comment (sharidas) [reply lib/Command/FixEncryptedVersion.php]: I didn't find anything in the code which could help us. I do understand the concern of relying on a very generic exception like `HintException` :(

File: lib/Command/FixEncryptedVersion.php [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+			\OC::$server->getLogger()->warning("Issue: " . $e->getMessage());
+			//If allowOnce is set to false, this becomes recursive.
+			if ($allowOnce === true) {
+				//Lets rectify the file by correcting encrypted version
+				return $this->correctEncryptedVersion($path, $output);
```
[the remark this answers is not in the document]
Comment (sharidas) [reply lib/Command/FixEncryptedVersion.php]: Done

File: lib/Command/FixEncryptedVersion.php [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+		$fixedFileInfo = ['filename' => '', 'encryptedVersion' => 0];
+
+		$fileInfo = $this->view->getFileInfo($path);
+		$fileId = $fileInfo->getId();
+		$encryptedVersion = $fileInfo->getEncryptedVersion();
+		$oldEncryptedVersion = $encryptedVersion;
+
+		$storage = $fileInfo->getStorage();
+
+		$cache = new Cache($storage);
```
[the remark this answers is not in the document]
Comment (sharidas) [reply lib/Command/FixEncryptedVersion.php]: Done. Thanks for the pointer.

File: lib/Command/FixEncryptedVersion.php [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+		$fileCache = $cache->get($fileId);
+		$fileCacheStorage = $fileCache->getData()['storage'];
```
[the remark this answers is not in the document]
Comment (sharidas) [reply lib/Command/FixEncryptedVersion.php]: No more relying on the string. Because in the updated version of pr I have used :
```
if ($storage->instanceOfStorage('OCA\Files_Sharing\ISharedStorage'))
```

File: lib/Command/FixEncryptedVersion.php [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+
+		if (\strpos($fileCacheStorage, "shared::") !== false) {
```
[the remark this answers is not in the document]
Comment (sharidas) [reply lib/Command/FixEncryptedVersion.php]: Done. Reference: https://github.com/owncloud/encryption/pull/115#discussion_r272562862

File: lib/Command/FixEncryptedVersion.php [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+			$output->writeln("<info>The file: $path is a share. Hence kindly fix this by running the script under the owner of share</info>");
+			return true;
+		}
+
+		if ($encryptedVersion >= 0) {
+			if ($encryptedVersion === 0) {
+				$skipDecrymentedVersion = true;
+			}
+			//test by decrementing the value till 1 and if nothing works try incrementing
```
[the remark this answers is not in the document]
Comment (sharidas) [reply lib/Command/FixEncryptedVersion.php]: Yah.

File: lib/Command/FixEncryptedVersion.php (hunk 1 above)
Comment (sharidas) [reply lib/Command/FixEncryptedVersion.php]: Done. Fixed with 5.

File: lib/Command/FixEncryptedVersion.php (hunk 2 above)
Comment (sharidas) [reply lib/Command/FixEncryptedVersion.php]: Removed this confusing logic. And made it look simple.

File: lib/Command/FixEncryptedVersion.php [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+			if (($skipDecrymentedVersion !== true) && ($fixedFileInfo['filename'] !== $path)) {
+				$encryptedVersion--;
+				while ($encryptedVersion > 0) {
```
[the remark this answers is not in the document]
Comment (sharidas) [reply lib/Command/FixEncryptedVersion.php]: Done. Removed `skipDecrymentedVersion`.

File: lib/Command/FixEncryptedVersion.php [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+					if ($fixedFileInfo['filename'] === $path && $fixedFileInfo['encryptedVersion'] > 0) {
```
[the remark this answers is not in the document]
Comment (sharidas) [reply lib/Command/FixEncryptedVersion.php]: Removed the `$filedFileInfo` usage. So no need to look into the filename and version, its already available in the loop.

File: lib/Command/FixEncryptedVersion.php (hunk 3 above)
Comment (sharidas) [reply lib/Command/FixEncryptedVersion.php]: Actually it doesn't work like that, I have added php comment. Also adjusted the variable `encryptedVersion` in this section to `$newEncryptedVersion`, the `oldEncryptedVersion` is renamed to `$wrongEncryptedVersion`.
So basically its like this:

when the $wrongEncryptedVersion = 6 at start

cycle 1: $increment=1, $newEncryptedVersion = 6+1 = 7
cycle 2: $increment=1, $newEncryptedVersion = 6+2 = 8
cycle 3: $increment=1, $newEncryptedVersion = 6+3 = 9
Basically the `$wrongEncryptedVersion` remains same and the `$increment` is doing the increment.

File: tests/unit/Command/FixEncryptedVersionTest.php (hunk 4 above)
Comment (sharidas) [reply tests/unit/Command/FixEncryptedVersionTest.php]: Added further increments in the same test.

File: lib/Command/FixEncryptedVersion.php (hunk 3 above)
Comment (PVince81) [reply lib/Command/FixEncryptedVersion.php]: so it was only a code readability issue, cool

Comment (sharidas) [reply lib/Command/FixEncryptedVersion.php]: Yup.

File: tests/unit/Command/FixEncryptedVersionTest.php [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+	/** @var CommandTester */
+	private $commandTester;
+
+	public static function setUpBeforeClass() {
+		parent::setUpBeforeClass();
+		\OC::$server->getUserManager()->createUser(self::TEST_ENCRYPTION_VERSION_AFFECTED_USER, 'foo');
+		//Enable encryption app
+		\OC_App::enable("encryption");
+		//Enable encryption
+		\OC::$server->getConfig()->setAppValue('core', 'encryption_enabled', 'yes');
```
Comment (PVince81) [inline tests/unit/Command/FixEncryptedVersionTest.php]: normally you should save appconfig values in a variable and set that variable back at the end of the test to avoid side effects

File: tests/unit/Command/FixEncryptedVersionTest.php [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+		$this->fixEncryptedVersion = new FixEncryptedVersion($this->rootFolder, $this->userManager, $this->view);
+		$this->commandTester = new CommandTester($this->fixEncryptedVersion);
+	}
+
+	/**
+	 * In this test the encrypted version is set to zero whereas it should have been
+	 * set to a positive non zero number.
+	 */
+	public function testEncryptedVersionZero() {
+		$this->markTestSkipped("Skipping this test because the user login would fail in the CI. And once we have a db endpoints available in acceptance test, we would add the test over there");
```
Comment (PVince81) [inline tests/unit/Command/FixEncryptedVersionTest.php]: please continue researching a bit

you should use `$this->loginAsUser()` for a more accurate login simulation. see https://github.com/owncloud/core/blob/v10.1.1/tests/lib/TestCase.php#L368

**Comment by sharidas** [comment]
The output buffer issue https://drone.owncloud.com/owncloud/encryption/609/150 :(

**Comment by sharidas** [comment]
Here is the output captured from my instance 
[a.txt](https://github.com/owncloud/encryption/files/3054520/a.txt) the output looks clean. @PVince81 let me know if this looks ok.

**Comment by PVince81** [comment]
> Here is the output captured from my instance

not sure what this is and what I should do about it ?

**Comment by sharidas** [comment]
> > Here is the output captured from my instance
> 
> not sure what this is and what I should do about it ?



> > Here is the output captured from my instance
> 
> not sure what this is and what I should do about it ?

Ok :) Lets wait for the CI result.

**Comment by sharidas** [comment]
Will have to restart the CI again, because of unrelated failure https://drone.owncloud.com/owncloud/encryption/612/1494 :(

**Comment by sharidas** [comment]
The CI has passed. Would require re-review.

**Comment by PVince81** [comment]
looks good now.

I'm missing a test where the correct value is lower than the encrypted version value and where decrementing would succeed.

**Comment by sharidas** [comment]
> I'm missing a test where the correct value is lower than the encrypted version value and where decrementing would succeed.

Ouch. Added the test at the end of the test file, which shows that the command would allow the files encrypted version reach the correct value by decrementing only.
